### ISSUE TITLE
Fl_Help_View::find returns strlen of value on success

### ISSUE THREAD

**Issue by MoAlyousef** [fltk/fltk#179]
Hello

I noticed an issue with Fl_Help_View, is that on success it always returns strlen of the inner value.

```cpp
#include <FL/Fl.H>
#include <FL/Fl_Help_View.H>
#include <FL/Fl_Window.H>

int main() {
    Fl_Window w(400, 300);
    Fl_Help_View v(0, 0, 400, 300);
    v.value("Hello <b>World</b>");
    printf("%d\n", v.find("whatever")); // prints -1 as it should
    printf("%d\n", v.find("Hello")); // prints 18, should print 0 I presume
    printf("%ld\n", strlen(v.value())); // prints 18
    w.end();
    w.show();
    return Fl::run();
}
```

**Comment by Albrecht-S** [comment]
Thanks for the report. Indeed the [documented behavior](https://www.fltk.org/doc-1.4/classFl__Help__View.html#a50ad770ff85426da7eb7859f6b747187) is:
> Returns the matching position or -1 if not found

I confirm that this is a bug and attached the label "bug".

**Comment by Albrecht-S** [comment]
OK, I've been working on this and I believe I found the bugs (three bugs in total in one function,  what a mess). And still one well-known `*FIXME* *UTF-8*` left (which can not be solved easily, but it's possible). Sigh.

I need some more time and tests and I'll post an update tomorrow.

**Comment by Albrecht-S** [comment]
FWIW, my current test results:
```
pos    = '01234567 10 234567 20 234567 30 234567 40 234567 50'
value  = 'Hello <b>World</b><p>Hello Bug: 1 &euro; (€)</p>'
length = 50
 find( whatever, 0) ->  -1
 find(    Hello, 0) ->   0
 find(       el, 0) ->   1
 find(       ll, 0) ->   2
 find(      llo, 0) ->   2
 find(      llo, 1) ->   2
 find(      llo, 2) ->   2
 find(      llo, 3) ->  23
 find(       lo, 0) ->   3
 find(    World, 0) ->   9
 find(    hello, 5) ->  21
 find(     ello, 5) ->  22
 find(      llo, 5) ->  23
 find(      Bug, 5) ->  27
 find(       ug, 5) ->  28
 find(      €, 5) ->  42
 find(   &euro;, 5) ->  -1
```
and the modified demo program [help_view.cxx.txt](https://github.com/fltk/fltk/files/5778922/help_view.cxx.txt) (attached).

And finally, for completeness, the preliminary patch [Fl_Help_View.diff.txt](https://github.com/fltk/fltk/files/5778941/Fl_Help_View.diff.txt) in case someone likes to test which I would appreciate. All feedback welcome.

Hint: try the new demo program first with an unpatched library to compare the results.

**Comment by MoAlyousef** [comment]
The output from the patch seems correct:
```
pos    = '01234567 10 234567 20 234567 30 234567 40 234567 50'
value  = 'Hello <b>World</b><p>Hello Bug: 1 &euro; (€)</p>'
length = 50
 find( whatever, 0) ->  -1
 find(    Hello, 0) ->   0
 find(       el, 0) ->   1
 find(       ll, 0) ->   2
 find(      llo, 0) ->   2
 find(      llo, 1) ->   2
 find(      llo, 2) ->   2
 find(      llo, 3) ->  23
 find(       lo, 0) ->   3
 find(    World, 0) ->   9
 find(    hello, 5) ->  21
 find(     ello, 5) ->  22
 find(      llo, 5) ->  23
 find(      Bug, 5) ->  27
 find(       ug, 5) ->  28
 find(      €, 5) ->  42
 find(   &euro;, 5) ->  -1
```

Another question is whether the html elements should be skipped or not? For example:
```
find(    World, 0) ->   9 // should this be 6 or 9
```

**Comment by Albrecht-S** [comment]
@MoAlyousef Thanks for testing and for your feedback
> Another question is whether the html elements should be skipped or not? For example:
> find(    World, 0) ->   9 // should this be 6 or 9

Good question, for two reasons:
1. I think the answer for your example is 9 because this is the offset of the search string in the HTML buffer (`value()`). That's what you'd want (usually, IMHO), and that's also consistent with the way the 2nd argument (`starting position p`) is used: as a start offset in the `value()` string.
2. But the same question can also be asked WRT **matching** strings, for instance: should the search string `Hello World` match HTML text like `Hello <b>World</b>` in value() ?

I'm tempted to say "yes" to the 2nd example, but what if the HTML is `Hello <br>World` ? In the first case you could ignore the `<b>` because it does not "change" the text (only formatting) but the second case would introduce a line break which would usually not match the search string in a plain text file.

That's, BTW, a real case I stumbled across: the search string `World` matched in the old code but would (after my first "simple" mods) not return 9 ("World") as expected but 6, i.e. it "matched" `<b>World`: it started the string comparison at position 6, skipped `<b>`, found a match, and returned the position of the _beginning of the string comparison_ which was 6 (not 9). My current fix is to reset the search (aka string comparison) whenever the search finds a tag (`<`) which is skipped, but I thought about changing this in the next iteration of my patch. Here's a more interesting example and comparison with the old code:

1. Old code would match in this example (returning '18'):
```
pos    = '01234567 10 234567 20 234567 30 234567 40 234567 50'
value  = 'Hello W<b>orld</b><p>Hello Bug: 1 &euro; (€)</p>'
length = 50
 find( whatever, 0) ->  -1
 find(Hello World, 0) ->  18
 find(    World, 0) ->  18
```

2. New code would not match:
```
pos    = '01234567 10 234567 20 234567 30 234567 40 234567 50'
value  = 'Hello W<b>orld</b><p>Hello Bug: 1 &euro; (€)</p>'
length = 50
 find( whatever, 0) ->  -1
 find(Hello World, 0) ->  -1
 find(    World, 0) ->  -1
```

So, the old code's intention was indeed to skip HTML tags but I'm not sure if that's the best solution. Note that I could "fix" the wrong return value w/o sacrificing to skip the HTML tags but again, I don't know what the correct/best solution would be.

After all, this method (find()) didn't work even in FLTK 1.1 (the code is basically the same) and the documentation doesn't specify how a "match" should be determined, hence we're IMHO free to implement whatever we like to decide now and fill the gaps in the documentation specifying what a "match" is and what position the function returns.

**Comment by Albrecht-S** [comment]
FWIW: Fl_Help_View::find() is not documented in [FLTK 1.1 docs](https://www.fltk.org/doc-1.1/Fl_Help_View.html).

> ... hence we're IMHO free to implement whatever we like to decide now and fill the gaps in the documentation ...

We could even add a 3rd argument `skip_tags` to let the user decide ... or something like that.

**Comment by MoAlyousef** [comment]
I think it’s reasonable to not skip the tags. Your example of using br tags makes sense. Html parsing is quite complex on it’s own and should probably be relegated to a 3rd party library if the user needs it.

**Comment by Albrecht-S** [comment]
I agree to skip the tags. The result is that HTML tags never match, i.e. they are treated as an existing character that does not match anything in the input string. The [updated documentation](https://fltk.gitlab.io/fltk/classFl__Help__View.html#a50ad770ff85426da7eb7859f6b747187) is now available on our GitLab mirror but not (yet) on fltk.org (will be updated later). Feedback welcome.

I committed the fix in git commit fba4d58509e337c3f868c7ae68b3566f85ba86f5, test results appreciated.

Thanks for helping to fix this issue.

**Comment by Albrecht-S** [comment]
@MoAlyousef Thanks for your :+1: and your help, I take it this issue is done and I'm closing it now.
